synapse/handlers/e2e_keys.py

```diff
@@ -976,16 +982,126 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
+            key is None
+            and not self.is_mine(user)
+            # We only get "master" and "self_signing" keys from remote servers
+            and key_type in ["master", "self_signing"]
+        ):
+            (
+                key,
+                key_id,
+                verify_key,
+            ) = yield self._retrieve_cross_signing_keys_for_remote_user(user, key_type)
+
         if key is None:
-            logger.debug("no %s key found for %s", key_type, user_id)
+            logger.warning("No %s key found for %s", key_type, user_id)
             raise NotFoundError("No %s key found for %s" % (key_type, user_id))
-        key_id, verify_key = get_verify_key_from_cross_signing_key(key)
+
+        try:
+            key_id, verify_key = get_verify_key_from_cross_signing_key(key)
+        except ValueError as e:
+            logger.warning(
+                "Invalid %s key retrieved: %s - %s %s", key_type, key, type(e), e,
+            )
+            raise SynapseError(
+                502, "Invalid %s key retrieved from remote server" % (key_type,)
+            )
+
         return key, key_id, verify_key
 
+    @defer.inlineCallbacks
+    def _retrieve_cross_signing_keys_for_remote_user(
+        self, user: UserID, desired_key_type: str,
+    ):
+        """Queries cross-signing keys for a remote user and saves them to the database
+
+        Only the key specified by `key_type` will be returned, while all retrieved keys
+        will be saved regardless
+
+        Args:
+            user: The user to query remote keys for
+            desired_key_type: The type of key to receive. One of "master", "self_signing"
+
+        Returns:
+            Deferred[Tuple[Optional[Dict], Optional[str], Optional[VerifyKey]]]: A tuple
+            of the retrieved key content, the key's ID and the matching VerifyKey.
+            If the key cannot be retrieved, all values in the tuple will instead be None.
+        """
+        try:
+            remote_result = yield self.federation.query_user_devices(
+                user.domain, user.to_string()
+            )
+        except Exception as e:
+            logger.warning(
+                "Unable to query %s for cross-signing keys of user %s: %s %s",
+                user.domain,
+                user.to_string(),
+                type(e),
+                e,
+            )
+            return None, None, None
+
+        # Process each of the retrieved cross-signing keys
+        final_key = None
+        final_key_id = None
+        final_verify_key = None
+        device_ids = []
+        for key_type in ["master", "self_signing"]:
+            key_content = remote_result.get(key_type + "_key")
+            if not key_content:
+                continue
+
+            # At the same time, store this key in the db for
+            # subsequent queries
+            yield self.store.set_e2e_cross_signing_key(
+                user.to_string(), key_type, key_content
+            )
```

Comment:
is it correct that we store it before we check that it can be parsed? (In short, is the data in the table supposed to have been validated or not? If not, we need to be careful when we extract it). What does the existing code do?

Reply:
Considering the above call from the database did not have a try/except, I believe we're operating on the assumption that the data in the database is valid.

Yes, we should validate this data in some way. Some level of such is done in `do_remote_query`:

https://github.com/matrix-org/synapse/blob/461f01ad43d7d572da626f637ddeb018cad9fcbb/synapse/handlers/e2e_keys.py#L236-L248

I'd hate to duplicate but that could be copied.

Reply:
Speaking of verification, one thing that's confusing me is that in the loop of `_retrieve_cross_signing_keys_for_remote_user`, `remote_result` should look something like this:

https://github.com/matrix-org/synapse/blob/fd8d154e6da6830b1cc828b7e6dfcddf53b9cad2/synapse/federation/transport/client.py#L404-L415

Thus `key_content` should look something like:

https://github.com/matrix-org/synapse/blob/fd8d154e6da6830b1cc828b7e6dfcddf53b9cad2/synapse/federation/transport/client.py#L407-L408

And thus we should need to extract the `key_contents` from that dictionary before we pass it to `get_verify_key_from_cross_signing_key`.

And yet this PR works, so something isn't lining up here.

Reply:
Already, so my docstring update is incorrect, as the thing I actually get back from `query_client_keys` is in the form:

```
{
  "user_id": "@test18:workerstest.amorgan.xyz",
  "stream_id": 213,
  "devices": [
    {
      "device_id": "MLYKYOAKEZ"
    },
    {
      "device_id": "NBICDLRVHV",
      "keys": {
        "algorithms": [
          "m.olm.v1.curve25519-aes-sha2",
          "m.megolm.v1.aes-sha2"
        ],
        "device_id": "NBICDLRVHV",
        "keys": {
          "curve25519:NBICDLRVHV": "E9ryKjylPorQIxyRA4FsJwi9d2bPECLvWKjgcuqmamU",
          "ed25519:NBICDLRVHV": "yk5AMkCyhogBCQt1CrVbFL9weRb/5iExmIDYd7JmBhE"
        },
        "signatures": {
          "@test18:workerstest.amorgan.xyz": {
            "ed25519:NBICDLRVHV": "JTAL+uIduDDJjdQneocTbjqjIDMljqDOq8bJ6ZgAYOQ01wMHrX4qRmQufQT9KAiYoXQ+b6UY+ZNEIFjyi1rjCg",
            "ed25519:cEYsFkRyHrfj1fLLgq4rA+JyvOwHda2mO092N4scL/E": "RQSm3aYjvWMWHx0GHZPRRKvj90TEdlqjOGXulXBUSDEuI2P64A7N/7/uSdx//BhqOgcmZSS/9o1Awlpb+KiiCg"
          }
        },
        "user_id": "@test18:workerstest.amorgan.xyz"
      },
      "device_display_name": "https://riot.im/develop/ via Firefox on Linux"
    }
  ],
  "master_key": {
    "user_id": "@test18:workerstest.amorgan.xyz",
    "usage": [
      "master"
    ],
    "keys": {
      "ed25519:RPg3wmdeZOvjMCmi3/pP9+4ndSW9W2dhE/RkLyCdr+E": "RPg3wmdeZOvjMCmi3/pP9+4ndSW9W2dhE/RkLyCdr+E"
    },
    "signatures": {
      "@test18:workerstest.amorgan.xyz": {
        "ed25519:NBICDLRVHV": "YLdKUD4aDmJx/vqIWjbqXTOZXv+cU1ba8Z9hw44vVn32uhLdJAiGC+L/6HeBTz7+Wh+NdWGBiy+usea3408bDA"
      }
    }
  },
  "self_signing_key": {
    "user_id": "@test18:workerstest.amorgan.xyz",
    "usage": [
      "self_signing"
    ],
    "keys": {
      "ed25519:cEYsFkRyHrfj1fLLgq4rA+JyvOwHda2mO092N4scL/E": "cEYsFkRyHrfj1fLLgq4rA+JyvOwHda2mO092N4scL/E"
    },
    "signatures": {
      "@test18:workerstest.amorgan.xyz": {
        "ed25519:RPg3wmdeZOvjMCmi3/pP9+4ndSW9W2dhE/RkLyCdr+E": "1H9aaL9623L764uGAezZM5jH3Gp5DLwto4s9rz+IP3bGQqLo0Cyi2bWK54aJbVrJ6akwxahZRs/n9vLDn3nIDQ"
      }
    }
  }
}
```

Lemme update that.

In terms of verification, I think we just need to check the user_id matches the user we're querying for, and that there's a `keys` field under each `*_key`.

Reply:
Aaand I was looking at the output of `query_user_devices`, not `query_client_keys`.

Reply:
So to conclude, we now validate the key by checking it belongs to the right user, and has a valid `keys` key dict in it with `get_verify_key_from_cross_signing_key`.